staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/bootstrap/bootstrap.go

```
@@ -0,0 +1,98 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Package bootstrap provides a token authenticator for TLS bootstrap secrets.
+*/
+package bootstrap
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/kubernetes/pkg/client/informers/informers_generated/internalversion/core/internalversion"
+	listers "k8s.io/kubernetes/pkg/client/listers/core/internalversion"
+)
+
+const (
+	SecretType = "bootstrap.kubernetes.io/token"
+
+	TokenID     = "token-id"
+	TokenSecret = "token-secret"
+
+	BootstrapUserPrefix = "system:bootstrap:"
+	BootstrapGroup      = "system:bootstrappers"
+)
+
+// NewTokenAuthenticator initializes a bootstrap token authenticator.
+func NewTokenAuthenticator(informer internalversion.SecretInformer, namespace string) *TokenAuthenticator {
+	return &TokenAuthenticator{informer.Lister().Secrets(namespace)}
+}
+
+// TokenAuthenticator authenticates bootstrap tokens from secrets in the API server.
+type TokenAuthenticator struct {
+	lister listers.SecretNamespaceLister
+}
+
+// AuthenticateToken tries to match the provided token to a bootstrap token secret
+// in the "kube-system" namespace. If found, it authenticates the token in the
+// "system:bootstrappers" group and with the "system:bootstrap:(token-id)" username.
+//
+// All secrets must be of type "bootstrap.kubernetes.io/token". An example secret:
+//
+//     apiVersion: v1
+//     kind: Secret
+//     metadata:
+//       name: bootstrap-token-( token id )
+//       namespace: kube-system
+//     data:
+//       token-secret: ( private part of token )
+//       token-id: ( token id )
+//     type: bootstrap.kubernetes.io/token
+//
+func (t *TokenAuthenticator) AuthenticateToken(token string) (user.Info, bool, error) {
```

first check the format of the token and fail fast if it's not in the expected format `x:x`